File: ps1/Lagrange.java
/*
 * Lagrange.java
 *
 * Author:          Sam Tymorek, samtymorek@hotmail.com
 * TF:              <your TF's name>
 * Date modified:   09/21/2010
 */

/**
 * An application that reads positive integers from the user and
 * breaks them into sums of at most four positive perfect squares.
 */

import java.util.*;

public class Lagrange {
    /* Put your data members below. */
    private int[] terms = new int[4];

    public static ArrayList<Integer> squaresLessThan = new ArrayList<Integer>();
      
    /**
     * largestSquare - returns the largest perfect square less than or
     * or equal to the specified positive integer.
     */
    private static int largestSquare(int n) {
        if (n <= 0)
            throw new IllegalArgumentException("n must be positive");
        int temp = (int)Math.sqrt(n);
        return temp * temp;
    }
    
    // TEMPorarily public
    private static void getSquaresLessThanNum(int n) {
        if (n < 0)
            throw new IllegalArgumentException("n must be positive");
        if (n < 4) {
            squaresLessThan.add(1);
            return;
        } else {
            int temp = largestSquare(n);
            squaresLessThan.add(temp);
            temp--;
            getSquaresLessThanNum(temp);
        }    
    }    

    /* Put your constructors and methods below. */  
    public int numTerms() {
        int num = 0;
        for (int i=0; i<terms.length; i++) {
            if(terms[i] > 0)
                num++;
        };
        return num;
    }

    public Lagrange() {
        if (squaresLessThan.size() > 0) {
            squaresLessThan.clear();
        }
    }    
    
    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        
        while(true) {
            System.out.print("Enter a positive integer (-1 to quit):");
            int num = in.nextInt(); 
            if (num == -1) {
                System.out.println("Goodbye!");
                break;
            }
            
            Lagrange lagrange = new Lagrange();
//            lagrange.findSum(num); 
           
            System.out.print(Integer.toString(num) + "=");
            for(int i = 0; i < lagrange.numTerms(); i++) {
                if (i != lagrange.numTerms() - 1) {
                    System.out.print(Integer.toString(lagrange.terms[i]) + "+");
                } else {   
                    System.out.println(Integer.toString(lagrange.terms[i]));
                }
            }
            lagrange.getSquaresLessThanNum(num);
            System.out.println(lagrange.squaresLessThan);
        }
    }

}

File: ps1/lagrange.py
import math

squares = []
terms = []

def largest_square(n):
	temp = int(math.sqrt(n))
	return temp * temp
	
def squares_less_than(n):
	# 1 is only perfect square lower than four, so just return 1 and quit
	if n < 4:
		squares.append(1)
		return True
		 
	else:
		temp = largest_square(n)
		squares.append(temp)
		temp -= 1
		squares_less_than(temp)

ORIG_NUM = 7
squares_less_than(ORIG_NUM)
#squares.append(1)		

def find_sum(n,i):

#	print n
	if sum(terms) == ORIG_NUM and i <= 4:
		print 'success! %s' % terms
		return True
		
	for val in squares[i:]:
		if len(terms) < 4:
			terms.append(val)
			if find_sum(n-squares[i], i):
#				print 'good terms so far: %s' % terms
				return True
			terms.remove(val)
#	print 'bad terms: %s' % terms	
	return False
	
find_sum(ORIG_NUM,0)
		
		
		
		
		
		
		
		
